### PR TITLE
Speed optimizations & enable Valgrind in CI

### PR DESCRIPTION
**Pull request by jkflying**
This PR was the result of some profiling work on a live flight. Some major bottlenecks were identified and dealt with in the first few commits.

After that there was some cleanup, including some uninitialized variables which caused tests to fail only if run in a certain order, so I've added Valgrind to the CI so we don't deal with these again.

### CODE REVIEW

File: local_planner/src/nodes/waypoint_generator.cpp [hunk 1]
```diff
@@ -143,9 +142,8 @@ void WaypointGenerator::backOff() {
   output_.goto_position = toPoint(toEigen(pose_.pose.position) + dir);
   output_.goto_position.z = planner_info_.back_off_start_point.z;
 
-  output_.position_waypoint = createPoseMsg(output_.goto_position, last_yaw_);
+  output_.position_waypoint = createPoseMsg(output_.goto_position, curr_yaw_);
```
Comment (baumanta) [inline local_planner/src/nodes/waypoint_generator.cpp]: This function is not really in use anymore, so I don't think it matters much. But I would say that last_yaw was a bit more correct, as it reduces drift in yaw orientation a bit

File: local_planner/src/nodes/planner_functions.cpp [hunk 2]
```diff
@@ -244,31 +245,69 @@ void getCostMatrix(const Histogram& histogram, const Eigen::Vector3f& goal,
...
   float other_costs = 0.f;
   // reset cost matrix to zero
   cost_matrix.resize(GRID_LENGTH_E, GRID_LENGTH_Z);
-  cost_matrix.fill(0.f);
+  cost_matrix.fill(NAN);
 
   // fill in cost matrix
   for (int e_index = 0; e_index < GRID_LENGTH_E; e_index++) {
-    for (int z_index = 0; z_index < GRID_LENGTH_Z; z_index++) {
+    int z_scale = static_cast<int>(std::round(
```
[the remark this answers is not in the document]
Comment (jkflying) [reply local_planner/src/nodes/planner_functions.cpp]: I'll add comments to better explain what the z_scale variable is doing

**Comment by LorenzMeier** [comment]
What is the performance improvement in percent or whatever metric you were using?

**Comment by jkflying** [comment]
It isn't real flight data, but based on the unit tests it looks like about 30% reduction in calls to `sinf`/`cosf` in total, and our smoothing costs half of what it did before.

image

**Pull request merged**